### PR TITLE
[FLINK-20114][connector/kafka,common] Fix a few issues reported in Kafka Source release testing.

### PR DESCRIPTION
**Pull request by becketqin**
# What is the purpose of the change
This PR fixes a few issues reported during Kafka Source release testing.
- Do not commit offsets for partitions whose offsets has not been initialized.
- Do not log and fail the job again if exception is received by the UncaughtExceptionHandler in SourceCoordinatorContext.
- Report numRecordsOut from the SourceOperator.
- Close admin client early when periodic partition discovery is disabled.
- Close the OperatorCoordinators synchronously to avoid ClassNotFoundException.

## Brief change log
- Do not commit offsets for partitions whose offsets has not been initialized.
- Do not log and fail the job again if exception is received by the UncaughtExceptionHandler in SourceCoordinatorContext.
- Report numRecordsOut from the SourceOperator.
- Close admin client early when periodic partition discovery is disabled.
- Close the OperatorCoordinators synchronously to avoid ClassNotFoundException.

## Verifying this change
Two unit tests have been added to verify the following two commits.
- Do not commit offsets for partitions whose offsets has not been initialized.
- Do not log and fail the job again if exception is received by the UncaughtExceptionHandler in SourceCoordinatorContext.

Verified via the WebUI that the metrics of records out is correctly reported.
Used the `StateMachineExample` to start/cancel/resume jobs and did not see dangling `AdminClients` / `ClassNotFoundException`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java [hunk 1]
```diff
@@ -170,6 +178,70 @@ public void testSnapshotAndRestore() throws Exception {
 
 	}
 
+	@Test
+	public void testCallableInterruptedDuringShutdownDoNotFailJob() throws InterruptedException {
```
Comment (StephanEwen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java:182]: This test looks to be testing against a specific implementation, rather against a contract.
You need to assume the exact specific way of how the `close()` method is implemented for this test to be meaningful. The test re-engineers how `close()` works to produce a special race.

That usually causes issues: as soon as the `close()` implementation changes
  - either the test breaks (despite everything being okay)
  - or the test becomes meaningless (doesn't reproduce the critical situation any more).

Testing such concurrent race situations is always a bit tricky. I think the "manual executors" have a decent trade-off between reliably producing situations and not assuming too much implementation.
You could take a look at the `ContinuousFileSplitEnumeratorTest` and the `ManuallyTriggeredScheduledExecutorService` in the `TestingSplitEnumeratorContext`.

Using that, you can probably replace all the complex code in the test by:
```java
// async action before the enumerator closes
testingContext.callAsync(actionThatThrowsException);
asyncExecutor.triggerAll();

// callback after the enumerator closes
testingContext.close();
enumeratorExecutor.triggerAll();

assertFalse(operatorCoordinatorContext.isJobFailed());
```

This only ties the test to the implementation aspect that there are two executors, not any specific structure of the close method.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 2]
```diff
@@ -321,4 +328,37 @@ private void registerReader() {
 	ListState<SplitT> getReaderState() {
 		return readerState;
 	}
+
+	// ----------- private class --------------
+
+	private static class CountingDataOutput<T> implements DataOutput<T> {
```
Comment (StephanEwen) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:334]: Can we put the counting in `SourceOperatorStreamTask`?
That way we can
  - avoid yet another wrapping layer for the output
  - have the record counting the responsibility of the task and operator chain, which looks like the level on which it is solved in the remaining cases

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java [hunk 3]
```diff
@@ -131,7 +131,14 @@ public void start() {
 			LOG.info("Starting the KafkaSourceEnumerator for consumer group {} " +
 				"without periodic partition discovery.", consumerGroupId);
 			context.callAsync(
-					this::discoverAndInitializePartitionSplit,
+					() -> {
+						try {
+							return discoverAndInitializePartitionSplit();
+						} finally {
+							// Close the admin client early because we won't use it anymore.
+							adminClient.close();
```
Comment (StephanEwen) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java:144]: Is the admin client created in the `start()` method and closed in the start method? Should it be a local variable then? Having it as a class field suggest other methods can use it.

Edit: Seems it is used in other places, which seems a bit hard to understand now. What is the life-cycle and use of that field?

**Review by StephanEwen — COMMENTED** [review]
A few comments for changes, otherwise looks good.

**Comment by rmetzger** [comment]
Thanks a lot for addressing the issues I've reported. While testing this PR, I noticed that the job got stuck while submitting it:

```
2020-11-26 20:46:43,453 INFO  org.apache.flink.runtime.dispatcher.StandaloneDispatcher     [] - Received JobGraph submission 362e7f6a2a9901e5d6d2ea69d40a69d4 (State machine job).
2020-11-26 20:46:43,453 INFO  org.apache.flink.runtime.dispatcher.StandaloneDispatcher     [] - Submitting job 362e7f6a2a9901e5d6d2ea69d40a69d4 (State machine job).
2020-11-26 20:46:43,455 INFO  org.apache.flink.runtime.rpc.akka.AkkaRpcService             [] - Starting RPC endpoint for org.apache.flink.runtime.jobmaster.JobMaster at akka://flink/user/rpc/jobmanager_4 .
2020-11-26 20:46:43,455 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Initializing job State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4).
2020-11-26 20:46:43,456 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Using restart back off time strategy FixedDelayRestartBackoffTimeStrategy(maxNumberRestartAttempts=2147483647, backoffTimeMS=1000) for State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4).
2020-11-26 20:46:43,457 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Running initialization on master for job State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4).
2020-11-26 20:46:43,457 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Successfully ran initialization on master in 0 ms.
2020-11-26 20:46:43,492 INFO  org.apache.flink.runtime.scheduler.adapter.DefaultExecutionTopology [] - Built 1 pipelined regions in 0 ms
2020-11-26 20:46:43,492 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - No state backend has been configured, using default (Memory / JobManager) MemoryStateBackend (data in heap memory / checkpoints to JobManager) (checkpoints: 'null', savepoints: 'null', asynchronous: TRUE, maxStateSize: 5242880)
2020-11-26 20:46:43,492 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator    [] - No checkpoint found during restore.
2020-11-26 20:46:43,493 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Using failover strategy org.apache.flink.runtime.executiongraph.failover.flip1.RestartPipelinedRegionFailoverStrategy@530fdf01 for State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4).
2020-11-26 20:46:43,493 INFO  org.apache.flink.runtime.jobmaster.JobManagerRunnerImpl      [] - JobManager runner for job State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4) was granted leadership with session id 00000000-0000-0000-0000-000000000000 at akka.tcp://flink@localhost:6123/user/rpc/jobmanager_4.
2020-11-26 20:46:43,493 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Starting execution of job State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4) under job master id 00000000000000000000000000000000.
2020-11-26 20:46:43,493 INFO  org.apache.flink.runtime.source.coordinator.SourceCoordinator [] - Starting split enumerator for source Source: Kafka Source.
2020-11-26 20:46:43,501 INFO  org.apache.flink.runtime.source.coordinator.SourceCoordinator [] - Closing SourceCoordinator for source Source: Kafka Source.
2020-11-26 20:46:43,502 INFO  org.apache.flink.runtime.source.coordinator.SourceCoordinator [] - Source coordinator for source Source: Kafka Source closed.
2020-11-26 20:46:43,502 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Connecting to ResourceManager akka.tcp://flink@localhost:6123/user/rpc/resourcemanager_*(00000000000000000000000000000000)
2020-11-26 20:46:43,503 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - Resolved ResourceManager address, beginning registration
2020-11-26 20:46:43,503 INFO  org.apache.flink.runtime.resourcemanager.StandaloneResourceManager [] - Registering job manager 00000000000000000000000000000000@akka.tcp://flink@localhost:6123/user/rpc/jobmanager_4 for job 362e7f6a2a9901e5d6d2ea69d40a69d4.
2020-11-26 20:46:43,504 INFO  org.apache.flink.runtime.resourcemanager.StandaloneResourceManager [] - Registered job manager 00000000000000000000000000000000@akka.tcp://flink@localhost:6123/user/rpc/jobmanager_4 for job 362e7f6a2a9901e5d6d2ea69d40a69d4.
2020-11-26 20:46:43,506 INFO  org.apache.flink.runtime.jobmaster.JobMaster                 [] - JobManager successfully registered at ResourceManager, leader id: 00000000000000000000000000000000.
---- manual cancellation of the job -----
2020-11-26 20:59:12,933 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph       [] - Job State machine job (362e7f6a2a9901e5d6d2ea69d40a69d4) switched from state CREATED to CANCELLING.
2020-11-26 20:59:12,933 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph       [] - Source: Kafka Source (1/4) (f08af83bdc18ed82549cafcc97b747a4) switched from CREATED to CANCELING.
```

I'm not sure if this problem is related to the scheduler or your changes, but it looks weird that the source coordinator got closed again right away.

**Comment by becketqin** [comment]
Hey @rmetzger , can you check if you are using the latest update of the PR? Some earlier commit I had for this PR had a problem of using `Integer.parseInt()` to parse a long value, which causes the `SplitEnumerator` to throw exception immediately. And it looks that the JM was not handling that correctly to fail the job. I have fired a ticket [FLINK-20382](https://issues.apache.org/jira/browse/FLINK-20382) for this.

**Review by becketqin — COMMENTED** [review]
@StephanEwen Thanks for the review and suggestion. I have updated the patch to address the comments.

**Comment by rmetzger** [comment]
I've rebuild Flink using the latest version of this PR + I added myself some logging into the JobMaster / Scheduler to find out the issue:

```
Caused by: org.apache.flink.util.FlinkRuntimeException: Failed to start the operator coordinators
        at org.apache.flink.runtime.scheduler.SchedulerBase.startAllOperatorCoordinators(SchedulerBase.java:1023) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.scheduler.SchedulerBase.startScheduling(SchedulerBase.java:490) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.jobmaster.JobMaster.startScheduling(JobMaster.java:952) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:705) ~[?:1.8.0_222]
        ... 27 more
Caused by: java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
        at org.apache.flink.connector.kafka.source.enumerator.KafkaSourceEnumerator.<init>(KafkaSourceEnumerator.java:116) ~[?:?]
        at org.apache.flink.connector.kafka.source.enumerator.KafkaSourceEnumerator.<init>(KafkaSourceEnumerator.java:91) ~[?:?]
        at org.apache.flink.connector.kafka.source.KafkaSource.createEnumerator(KafkaSource.java:135) ~[?:?]
        at org.apache.flink.runtime.source.coordinator.SourceCoordinator.start(SourceCoordinator.java:119) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$DeferrableCoordinator.start(RecreateOnResetOperatorCoordinator.java:366) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator.start(RecreateOnResetOperatorCoordinator.java:69) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.operators.coordination.OperatorCoordinatorHolder.start(OperatorCoordinatorHolder.java:182) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.scheduler.SchedulerBase.startAllOperatorCoordinators(SchedulerBase.java:1017) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.scheduler.SchedulerBase.startScheduling(SchedulerBase.java:490) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.runtime.jobmaster.JobMaster.startScheduling(JobMaster.java:952) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:705) ~[?:1.8.0_222]
        ... 27 more
```.

**Comment by rmetzger** [comment]
Note: CI is failing because of checkstyle.

**Comment by becketqin** [comment]
The patch migrates all the existing IT cases from `KafkaConsumerTestBase` for the new KafkaSource, except the `runEndOfStreamTest()` which is no longer applicable.

In order to migrate the IT case, the following two commits has to be cherry-picked from FLINK-20379.
- Add a `GenericDeserializationSchema` that suits the new Source API. The `GenericDeserializationSchema` is intended to take care of the deserialization from an arbitrary record type, which is usually a `byte[]` with additional information.
- Let the KafkaRecordDeserializer implement `GenericDeserializationSchema`. Add bridge methods to the existing `KafkaDeserializationSchema` implementations.

Other than that, the following two commits are made:
03173a7 fixes a bug where NPE was thrown when the committed offset does not exist for a partition.
e1fe5fd adjusts the behavior of `SpecifiedOffsetsInitializer` to keep it the same as the legacy `FlinkKafkaConsumer`.
df413b4 Disable offsets auto commit by default.

I have verified the patch with a modified `StateMachineExample` with the following steps:
1. Start a Kafka cluster and create a topic of 32 partitions.
2. Start a Flink standalone cluster with 3 TMs.
3. Submit the `KafkaEventsGeneratorJob` to generate records.
4. Submit the `StateMachineExample` to process the records.
  - The jobs runs without exception.
  - Checkpoint works fine.
  - The offsets are committed back to Flink correctly.
  - The WebUI shows the metrics correctly.
5. Kill -9 the TM that runs the `StateMachineExample`.
  - The Task fails over correctly and resumes consumption from the last checkpointed offset.
6. Cancel the `StateMachineExample` with a savepoint.
  - The job stops without an exception.
7. Resume the `StateMachineExample` from the savepoint.
  - The job resumes correctly from the offsets when it was canceled.
  - NOTE: The following exception may be seen from the TM log. It is a know issue from Kafka because the same client-id was registered twice to the JMX. It is a known issue to Kafka, but I am not familiar with the mbeanserver enough to tell if Flink should expect this.
```
javax.management.InstanceAlreadyExistsException: kafka.consumer:type=app-info,id=myGroup-0
        at com.sun.jmx.mbeanserver.Repository.addMBean(Repository.java:437) ~[?:1.8.0_172]
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerWithRepository(DefaultMBeanServerInterceptor.java:1898) ~[?:1.8.0_172]
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerDynamicMBean(DefaultMBeanServerInterceptor.java:966) ~[?:1.8.0_172]
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerObject(DefaultMBeanServerInterceptor.java:900) ~[?:1.8.0_172]
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerMBean(DefaultMBeanServerInterceptor.java:324) ~[?:1.8.0_172]
        at com.sun.jmx.mbeanserver.JmxMBeanServer.registerMBean(JmxMBeanServer.java:522) ~[?:1.8.0_172]
        at org.apache.kafka.common.utils.AppInfoParser.registerAppInfo(AppInfoParser.java:64) ~[blob_p-7527eaac8696e2e6d749b6e93038622aad2a03ee-2deed2871e614394bdf6e
692044a2daf:1.12-SNAPSHOT]
        at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:814) ~[blob_p-7527eaac8696e2e6d749b6e93038622aad2a03ee-2deed2871e614394bdf6e6920
44a2daf:1.12-SNAPSHOT]
        at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:666) ~[blob_p-7527eaac8696e2e6d749b6e93038622aad2a03ee-2deed2871e614394bdf6e6920
44a2daf:1.12-SNAPSHOT]
        at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:646) ~[blob_p-7527eaac8696e2e6d749b6e93038622aad2a03ee-2deed2871e614394bdf6e6920
44a2daf:1.12-SNAPSHOT]
        at org.apache.flink.connector.kafka.source.reader.KafkaPartitionSplitReader.<init>(KafkaPartitionSplitReader.java:84) ~[blob_p-7527eaac8696e2e6d749b6e9303862
2aad2a03ee-2deed2871e614394bdf6e692044a2daf:1.12-SNAPSHOT]
        at org.apache.flink.connector.kafka.source.KafkaSource.lambda$createReader$0(KafkaSource.java:121) ~[blob_p-7527eaac8696e2e6d749b6e93038622aad2a03ee-2deed287
1e614394bdf6e692044a2daf:1.12-SNAPSHOT]
        at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcherManager.createSplitFetcher(SplitFetcherManager.java:129) ~[flink-table-blink_2.11-1.12-S
NAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.connector.base.source.reader.fetcher.SingleThreadFetcherManager.addSplits(SingleThreadFetcherManager.java:59) ~[flink-table-blink_2.11-1.
12-SNAPSHOT.jar:1.12-SNAPSHOT]
        at org.apache.flink.connector.base.source.reader.SourceReaderBase.addSplits(SourceReaderBase.java:218) ~[flink-table-blink_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSH
OT]
        at org.apache.flink.streaming.api.operators.SourceOperator.open(SourceOperator.java:220) ~[flink-dist_2.11-1.12-SNAPSHOT.jar:1.12-SNAPSHOT]
```

**Comment by tweise** [comment]
Closing this PR since https://github.com/apache/flink/pull/15161 was merged.

**Pull request closed without merging**